File: InternetMap/MapData.m
//
//  MapData
//  InternetMap
//

#import "MapData.h"
#import "MapDisplay.h"
#import "Node.h"
#import "Lines.h"
#import "Connection.h"
#import "IndexBox.h"

@interface MapData ()
@end

@implementation MapData

-(Node*)nodeAtIndex:(NSUInteger)index {
    return [self.nodes objectAtIndex:index];
}

-(void)loadFromFile:(NSString*)filename {
    NSTimeInterval start = [NSDate timeIntervalSinceReferenceDate];
    
    NSString *fileContents = [NSString stringWithContentsOfFile:filename encoding:NSUTF8StringEncoding error:NULL];
    NSArray *lines = [fileContents componentsSeparatedByString:@"\n"];
    
    NSArray* header = [[lines objectAtIndex:0] componentsSeparatedByString:@"  "];
    int numNodes = [[header objectAtIndex:0] intValue];
    int numConnections = [[header objectAtIndex:1] intValue];
    
    self.nodes = [[NSMutableArray alloc] initWithCapacity:numNodes];
    self.nodesByAsn = [[NSMutableDictionary alloc] initWithCapacity:numNodes];

    for (int i = 0; i < numNodes; i++) {
        NSArray* nodeDesc = [[lines objectAtIndex:1 + i] componentsSeparatedByString:@" "];
        
        Node* node = [Node new];
        node.asn = [nodeDesc objectAtIndex:0];
        node.index = i;
        node.importance = [[nodeDesc objectAtIndex:1] floatValue];
        node.positionX = [[nodeDesc objectAtIndex:2] floatValue];
        node.positionY = [[nodeDesc objectAtIndex:3] floatValue];
        node.type = AS_UNKNOWN;
        
        [self.nodes addObject:node];
        [self.nodesByAsn setObject:node forKey:node.asn];
    }
    
    self.connections = [NSMutableArray new];
    
    for (int i = 0; i < numConnections; i++) {
        NSArray* connectionDesc = [[lines objectAtIndex:1 + numNodes + i] componentsSeparatedByString:@" "];
        
        Connection* connection = [Connection new];
        connection.first = [self.nodesByAsn valueForKey:[connectionDesc objectAtIndex:0]];
        connection.second = [self.nodesByAsn valueForKey:[connectionDesc objectAtIndex:1]];
        [self.connections addObject:connection];
    }
    
    [self createNodeBoxes];
    
    NSLog(@"load : %.2fms", ([NSDate timeIntervalSinceReferenceDate] - start) * 1000.0f);
}

- (void)createNodeBoxes {
    
    self.boxesForNodes = [NSMutableArray array];
    
    for (int k = 0; k < numberOfCellsZ; k++) {
        float z = IndexBoxMinZ + boxSizeZWithoutOverlap*k;
        for (int j = 0; j < numberOfCellsY; j++) {
            float y = IndexBoxMinY + boxSizeYWithoutOverlap*j;
            for(int i = 0; i < numberOfCellsX; i++) {
                float x = IndexBoxMinX + boxSizeXWithoutOverlap*i;
                IndexBox* box = [[IndexBox alloc] init];
                box.center = GLKVector3Make(x+boxSizeXWithoutOverlap/2, y+boxSizeYWithoutOverlap/2, z+boxSizeZWithoutOverlap/2);
                
                [self.boxesForNodes addObject:box];
            }
        }
    }
    
    for (int i = 0; i < [self.nodes count]; i++) {
        Node* node = [self.nodes objectAtIndex:i];
        GLKVector3 pos = [self.visualization nodePosition:node];
        IndexBox* box = [self indexBoxForPoint:pos];
        [box.indices addIndex:i];
    }
}

- (IndexBox*)indexBoxForPoint:(GLKVector3)point {
    GLKVector3 pos = point;
    
    //assumes all boxes have the same size
    float x = pos.x;
    if (x >= 0) {
        x += fabsf(IndexBoxMinX);
    }
    
    float y = pos.y;
    if (y >= 0) {
        y += fabsf(IndexBoxMinY);
    }
    
    float z = pos.z;
    if (z >= 0) {
        z += fabsf(IndexBoxMinZ);
    }
    
    int posX = (int)fabsf(x/boxSizeXWithoutOverlap);
    int posY = (int)fabsf(y/boxSizeYWithoutOverlap);
    int posZ = (int)fabsf(z/boxSizeZWithoutOverlap);
    int posInArray = posX + (fabsf(IndexBoxMinX) + fabsf(IndexBoxMaxX))/boxSizeXWithoutOverlap*posY + (fabsf(IndexBoxMinX) + fabsf(IndexBoxMaxX))/boxSizeXWithoutOverlap*(fabsf(IndexBoxMinY)+fabsf(IndexBoxMaxY))/boxSizeYWithoutOverlap*posZ;
    
    return [self.boxesForNodes objectAtIndex:posInArray];

}

- (void)addNodesToBox:(IndexBox*)box {
    for (int i = 0; i < [self.nodes count]; i++) {
        Node* node = [self.nodes objectAtIndex:i];
        GLKVector3 pos = [self.visualization nodePosition:node];
        if ([box isPointInside:pos]) {
            [box.indices addIndex:i];
        }
    }
}

-(void)loadFromAttrFile:(NSString*)filename {
    NSTimeInterval start = [NSDate timeIntervalSinceReferenceDate];

    NSDictionary *asTypeDict = [NSDictionary dictionaryWithObjectsAndKeys:
                                [NSNumber numberWithInt:AS_UNKNOWN], @"abstained",
                                [NSNumber numberWithInt:AS_T1], @"t1",
                                [NSNumber numberWithInt:AS_T2], @"t2",
                                [NSNumber numberWithInt:AS_COMP], @"comp",
                                [NSNumber numberWithInt:AS_EDU], @"edu",
                                [NSNumber numberWithInt:AS_IX], @"ix",
                                [NSNumber numberWithInt:AS_NIC], @"nic",
                                nil];

    NSString *fileContents = [NSString stringWithContentsOfFile:filename encoding:NSUTF8StringEncoding error:NULL];
    NSArray *lines = [fileContents componentsSeparatedByString:@"\n"];

    for(NSString *line in lines) {
        NSArray* asDesc = [line componentsSeparatedByString:@"\t"];
        
        Node* node = [self.nodesByAsn objectForKey:[asDesc objectAtIndex:0]];
        if(node){
            node.type = [[asTypeDict objectForKey: [asDesc objectAtIndex:7]] intValue];
            node.typeString = [asDesc objectAtIndex:7];
            node.textDescription = [asDesc objectAtIndex:1];
        }
    }
    
    
    NSLog(@"attr load : %.2fms", ([NSDate timeIntervalSinceReferenceDate] - start) * 1000.0f);
}


-(void)loadAsInfo:(NSString*)filename {
    NSTimeInterval start = [NSDate timeIntervalSinceReferenceDate];
    
    NSString *fileContents = [NSString stringWithContentsOfFile:filename encoding:NSUTF8StringEncoding error:NULL];
    NSError *parseError = nil;
    NSData* data = [fileContents dataUsingEncoding:NSUTF8StringEncoding];
    id jsonObject = [NSJSONSerialization JSONObjectWithData:data options:0 error:&parseError];
    
//    NSLog(@"%d", [jsonObject count]);
    for(id key in jsonObject)
    {
        Node* node = [self.nodesByAsn objectForKey:key];
        if(node){
            NSArray *as = [jsonObject objectForKey:key];
            node.name = [as objectAtIndex:1];
            node.textDescription = [as objectAtIndex:5];
            node.dateRegistered = [as objectAtIndex:3];
            node.address = [as objectAtIndex:7];
            node.city = [as objectAtIndex:8];
            node.state = [as objectAtIndex:9];
            node.postalCode = [as objectAtIndex:10];
            node.country = [as objectAtIndex:11];
        }
    }

    NSLog(@"asinfo load : %.2fms", ([NSDate timeIntervalSinceReferenceDate] - start) * 1000.0f);
}


-(void)updateDisplay:(MapDisplay*)display {
    NSTimeInterval start = [NSDate timeIntervalSinceReferenceDate];
    display.numNodes = self.nodes.count;
    [self.visualization updateDisplay:display forNodes:self.nodes];
    [self.visualization updateLineDisplay:display forConnections:self.connections];
        
    NSLog(@"update display : %.2fms", ([NSDate timeIntervalSinceReferenceDate] - start) * 1000.0f);
}

@end

File: InternetMap/MapData.cpp
//
//  MapData
//  InternetMap
//

#include "MapData.hpp"
#include "Node.hpp"
#include "DisplayLines.hpp"
#include "Connection.hpp"
#include "IndexBox.hpp"
#include "MapDisplay.hpp"
#include <sstream>
#include <stdlib.h>

// TODO: figure out how to do this right
#ifdef ANDROID
#include "ExternalCode/jsoncpp/json.h"
#else
#include "json.h"
#endif

NodePointer MapData::nodeAtIndex(unsigned int index) {
    return nodes[index];
}

void MapData::clear(void) {
    nodes .erase(nodes.begin(), nodes.end());
    connections.erase(connections.begin(), connections.end());
    boxesForNodes.erase(boxesForNodes.begin(), boxesForNodes.end());
    nodesByAsn.erase(nodesByAsn.begin(), nodesByAsn.end());
}

void split( std::vector<std::string> & theStringVector,  /* Altered/returned value */
      const  std::string  & theString,
      const  std::string  & theDelimiter)
{
    size_t  start = 0, end = 0;
    
    while ( end != std::string::npos)
    {
        end = theString.find( theDelimiter, start);
        
        // If at end, use length=maxLength.  Else use length=end-start.
        theStringVector.push_back( theString.substr( start,
                                                    (end == std::string::npos) ? std::string::npos : end - start));
        
        // If at end, use start=maxSize.  Else use start=end+delimiter.
        start = (   ( end > (std::string::npos - theDelimiter.size()) )
                 ?  std::string::npos  :  end + theDelimiter.size());
    }
}

static const int MAX_TOKEN_SIZE = 256;

const char* nextToken(const char* source, char* token, bool* lineEnd) {
    *lineEnd = false;
    
    while ((*source != ' ') && (*source != '\n') && (*source != 0)) {
        *token++ = *source++;
    }
    
    *lineEnd |= *source == '\n';
    
    *token = 0;

    if(*source != 0) {
        while ((*source == ' ') || (*source == '\n')) {
            source++;
            *lineEnd |= *source == '\n';
        }
    }
    
    return source;
}

void MapData::loadFromString(const std::string& text) {
    const char* sourceText = text.c_str();
    char token[MAX_TOKEN_SIZE];
    bool lineEnd;
    int numNodes, numConnections;

    sourceText = nextToken(sourceText, token, &lineEnd);
    numNodes = atof(token);

    sourceText = nextToken(sourceText, token, &lineEnd);
    numConnections = atof(token);
    
    nodes.reserve(numNodes);
    connections.reserve(numConnections);
    
    for (int i = 0; i < numNodes; i++) {
        NodePointer node(new Node());
        
        sourceText = nextToken(sourceText, token, &lineEnd);
        node->asn = token;
        node->index = i;
        sourceText = nextToken(sourceText, token, &lineEnd);
        node->importance = atof(token);
        sourceText = nextToken(sourceText, token, &lineEnd);
        node->positionX = atof(token);
        sourceText = nextToken(sourceText, token, &lineEnd);
        node->positionY = atof(token);
        node->type = AS_UNKNOWN;
        
        nodes.push_back(node);
        nodesByAsn[node->asn] = node;
    }

    for (int i = 0; i < numConnections; i++) {
        ConnectionPointer connection(new Connection());
        
        sourceText = nextToken(sourceText, token, &lineEnd);
        connection->first = nodesByAsn[token];
        sourceText = nextToken(sourceText, token, &lineEnd);
        connection->second = nodesByAsn[token];
        connection->first->connections.push_back(connection);
        connection->second->connections.push_back(connection);
        connections.push_back(connection);
    }

    createNodeBoxes();
    
    LOG("Loaded %d nodes, %d connections", numNodes, numConnections);
//    NSLog(@"load : %.2fms", ([NSDate timeIntervalSinceReferenceDate] - start) * 1000.0f);
}

void MapData::loadFromAttrString(const std::string& json){
    std::map<std::string, int> asTypeDict;
    asTypeDict["abstained"] = AS_UNKNOWN;
    asTypeDict["t1"] = AS_T1;
    asTypeDict["t2"] = AS_T2;
    asTypeDict["comp"] = AS_COMP;
    asTypeDict["edu"] = AS_EDU;
    asTypeDict["ix"] = AS_IX;
    asTypeDict["nic"] = AS_NIC;
    
    std::vector<std::string> lines;
    split(lines, json, "\n");

    for(unsigned int i = 0; i < lines.size(); i++) {
        
        std::string line = lines[i];
        std::vector<std::string> aDesc;
        split(aDesc, line, "\t");
        NodePointer node = nodesByAsn[aDesc[0]];
        
        if(node){
            
            node->type = asTypeDict[aDesc[7]];
            node->typeString = aDesc[7];
            node->textDescription = aDesc[1];
        }
    }

//    NSLog(@"attr load : %.2fms", ([NSDate timeIntervalSinceReferenceDate] - start) * 1000.0f);

}

void MapData::loadASInfo(const std::string& json){
    
    Json::Value root;
    Json::Reader reader;
    bool success = reader.parse(json, root);
    if(success) {
        std::vector<std::string> members = root.getMemberNames();
        for (unsigned int i = 0; i < members.size(); i++) {
            NodePointer node = nodesByAsn[members[i]];
            if (node) {
                Json::Value as = root[members[i]];
                node->name = as[1].asString();
                node->textDescription = as[5].asString();
                node->dateRegistered = as[3].asString();
                node->address = as[7].asString();
                node->city = as[8].asString();
                node->state = as[9].asString();
                node->postalCode = as[10].asString();
                node->country = as[11].asString();
            }
        }
    }
}

void MapData::updateDisplay(shared_ptr<MapDisplay> display){


//    NSTimeInterval start = [NSDate timeIntervalSinceReferenceDate];
    visualization->resetDisplayForNodes(display, nodes);
    visualization->updateLineDisplay(display, connections);
//    NSLog(@"update display : %.2fms", ([NSDate timeIntervalSinceReferenceDate] - start) * 1000.0f);

}


void MapData::createNodeBoxes() {
    
    for (int k = 0; k < numberOfCellsZ; k++) {
        float z = IndexBoxMinZ + boxSizeZWithoutOverlap*k;
        for (int j = 0; j < numberOfCellsY; j++) {
            float y = IndexBoxMinY + boxSizeYWithoutOverlap*j;
            for(int i = 0; i < numberOfCellsX; i++) {
                float x = IndexBoxMinX + boxSizeXWithoutOverlap*i;
                IndexBoxPointer box = IndexBoxPointer(new IndexBox());
                box->setCenter(Point3(x+boxSizeXWithoutOverlap/2, y+boxSizeYWithoutOverlap/2, z+boxSizeZWithoutOverlap/2));
                box->setMinCorner(Point3(x, y, z));
                box->setMaxCorner(Point3(x+boxSizeXWithoutOverlap, y+boxSizeYWithoutOverlap, z+boxSizeZWithoutOverlap));
                boxesForNodes.push_back(box);
            }
        }
    }
    
    for (unsigned int i = 0; i < nodes.size(); i++) {
        NodePointer ptrNode = nodes.at(i);
        Point3 pos = visualization->nodePosition(ptrNode);
        IndexBoxPointer box = indexBoxForPoint(pos);
        box->indices.insert(i);
    }
}

IndexBoxPointer MapData::indexBoxForPoint(const Point3& point) {
    
    int posX = (int)fabsf((point.getX() + fabsf(IndexBoxMinX))/boxSizeXWithoutOverlap);
    int posY = (int)fabsf((point.getY() + fabsf(IndexBoxMinY))/boxSizeYWithoutOverlap);
    int posZ = (int)fabsf((point.getZ() + fabsf(IndexBoxMinZ))/boxSizeZWithoutOverlap);
    int posInArray = posX + (fabsf(IndexBoxMinX) + fabsf(IndexBoxMaxX))/boxSizeXWithoutOverlap*posY + (fabsf(IndexBoxMinX) + fabsf(IndexBoxMaxX))/boxSizeXWithoutOverlap*(fabsf(IndexBoxMinY)+fabsf(IndexBoxMaxY))/boxSizeYWithoutOverlap*posZ;
    
    return boxesForNodes[posInArray];
}

